pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -57,20 +70,203 @@ type defaultStatefulSetControl struct {
+	selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	return ssc.controllerHistory.ListControllerRevisions(set, selector)
+}
+
+func (ssc *defaultStatefulSetControl) AdoptOrphanRevisions(
+	set *apps.StatefulSet,
+	revisions []*apps.ControllerRevision) error {
+	for i := range revisions {
+		adopted, err := ssc.controllerHistory.AdoptControllerRevision(set, controllerKind, revisions[i])
+		if err != nil {
+			return err
+		}
+		revisions[i] = adopted
+	}
+	return nil
+}
+
+// truncateHistory truncates any non-live ControllerRevisions in revisions from set's history. The UpdateRevision and
+// CurrentRevision in set's Status are considered to be live, as are any revisions associated with the Pods in pods.
+// Non-live revisions are deleted, starting with the revision with the lowest Revision, until only RevisionHistoryLimit
+// revisions remain. If the returned error is nil the operation was successful.
+func (ssc *defaultStatefulSetControl) truncateHistory(
+	set *apps.StatefulSet,
+	pods []*v1.Pod,
+	revisions []*apps.ControllerRevision,
+	current *apps.ControllerRevision,
+	update *apps.ControllerRevision) error {
+	history := make([]*apps.ControllerRevision, 0, len(revisions))
+	// mark all live revisions
+	live := map[string]bool{current.Name: true, update.Name: true}
+	for i := range pods {
+		live[getPodRevision(pods[i])] = true
+	}
+	// collect live revisions and historic revisions
+	for i := range revisions {
+		if !live[revisions[i].Name] {
+			history = append(history, revisions[i])
+		}
+	}
+	if len(history) < int(*set.Spec.RevisionHistoryLimit) {
+		return nil
+	}
+	// delete any non-live history to maintain the revision limit.
+	history = history[:int(*set.Spec.RevisionHistoryLimit)]
+	for i := 0; i < len(history); i++ {
+		if err := ssc.controllerHistory.DeleteControllerRevision(history[i]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getSetRevisions returns the current and update ControllerRevisions for set. This method may create a new revision,
+// or modify the Revision of an existing revision if an update to set is detected.
+func (ssc *defaultStatefulSetControl) getSetRevisions(
+	set *apps.StatefulSet,
+	revisions []*apps.ControllerRevision) (*apps.ControllerRevision, *apps.ControllerRevision, error) {
+	var currentRevision, updateRevision *apps.ControllerRevision
+
+	revisionCount := len(revisions)
+	history.SortControllerRevisions(revisions)
+
+	// create a new revision from the current set
+	updateRevision, err := newRevision(set, nextRevision(revisions))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// find any equivalent revisions
+	equalRevisions := history.FindEqualRevisions(revisions, updateRevision)
+	equalCount := len(equalRevisions)
+
+	if equalCount > 0 && history.EqualRevision(revisions[revisionCount-1], equalRevisions[equalCount-1]) {
```

Comment:
it is sorted by I will add a comment about the methods expectation of the supplied parameter.